### ISSUE TITLE
Idempotent with async messages

### ISSUE THREAD

**Issue by vykimo** [tulios/kafkajs#353]
I've discovered an unexpected behavior.
This small code will illustrate it well :
Here is my producer setup :
```
const producer = kafka.producer({
    maxInFlightRequests: 1,
    idempotent: true,   
    allowAutoTopicCreation: true
})
```
When I produce 10 asynchronous messages like this :
```
for (let i = 0; i < 10; i++)
        producer.send({
            topic: 'topic-test',
            messages: [
                { value: 'Hello KafkaJS user!'+ i },
            ],
        })
```

Here is the result on my consumer side (`numPartitions: 4`) :
```
{ partition: 1, offset: '37', value: 'Hello KafkaJS user!0' }
{ partition: 0, offset: '38', value: 'Hello KafkaJS user!9' }
{ partition: 2, offset: '37', value: 'Hello KafkaJS user!6' }
{ partition: 3, offset: '40', value: 'Hello KafkaJS user!7' }
```
Only random messages are coming and only 1 in each partition.
Are other messages lost?

In order to compare, when I set `idempotent: false,` or if I send sync messages (`await producer.send()`), I get the correct result : 
```
{ partition: 0, offset: '39', value: 'Hello KafkaJS user!0' }
{ partition: 2, offset: '38', value: 'Hello KafkaJS user!1' }
{ partition: 3, offset: '41', value: 'Hello KafkaJS user!2' }
{ partition: 0, offset: '40', value: 'Hello KafkaJS user!4' }
{ partition: 2, offset: '39', value: 'Hello KafkaJS user!5' }
{ partition: 3, offset: '42', value: 'Hello KafkaJS user!6' }
{ partition: 1, offset: '39', value: 'Hello KafkaJS user!3' }
{ partition: 1, offset: '40', value: 'Hello KafkaJS user!7' }
{ partition: 0, offset: '41', value: 'Hello KafkaJS user!8' }
{ partition: 2, offset: '40', value: 'Hello KafkaJS user!9' }
```

Please give me your thoughts.

**Comment by Nevon** [comment]
I have not been able to reproduce your issue. I wrote a test in b0b40f9 that attempts to replicate it. From just looking through your example, I don't see any reason it wouldn't work, although I wouldn't advocate for calling `producer.send` without handling the promise in any way (if it rejects, it'll crash your application). In my test, I'm sending the messages within `Promise.all`, and await that promise, but I also tried it your way with a simple for loop and not awaiting anything, and it seemed to work fine as well.

Could you share the rest of the context of the program? I would be interested in the consumer side of things, for example.

**Comment by vykimo** [comment]
It doesn't crash because I `try {} catch(e)` the whole block.

here is my simple consumer :
```
const consumer = kafka.consumer({
    allowAutoTopicCreation: true,
    groupId: 'test-group'
})
    await consumer.connect()
    await consumer.subscribe({ topic: 'topic-test', fromBeginning: true })

    await consumer.run({
        partitionsConsumedConcurrently: 4,
        eachMessage: async ({ topic, partition, message }) => {
            console.log({
                partition,
                offset: message.offset,
                value: message.value.toString(),
            })
        },
    })
```

**Comment by Nevon** [comment]
> It doesn't crash because I try {} catch(e) the whole block.

If you don't `await` the promise, it won't get caught by your try-catch. It'll be an unhandled promise rejection. Can you try adding a `.catch` to the `producer.send` call and log any error?

Could you also run your producer with log level `DEBUG` and share the logs? https://kafka.js.org/docs/configuration#logging

**Comment by vykimo** [comment]
Here is my producer logs : 
https://pastebin.com/EQ2rxrYE

**Comment by Nevon** [comment]
Looking at your debug log, I can see 11 successful produce requests in there, spread across all 4 partitions with no errors:

```
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":2,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":3,"errorCode":0,"baseOffset":"62","logAppendTime":"-1","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":3,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":1,"errorCode":0,"baseOffset":"62","logAppendTime":"-1","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":12,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":0,"errorCode":0,"baseOffset":"59","logAppendTime":"-1","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":11,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":1,"errorCode":0,"baseOffset":"62","logAppendTime":"1556623780829","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":10,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":3,"errorCode":0,"baseOffset":"63","logAppendTime":"-1","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":9,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":2,"errorCode":0,"baseOffset":"59","logAppendTime":"-1","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":8,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":0,"errorCode":0,"baseOffset":"59","logAppendTime":"1556623780834","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":7,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":1,"errorCode":0,"baseOffset":"62","logAppendTime":"1556623780829","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":6,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":3,"errorCode":0,"baseOffset":"63","logAppendTime":"1556623780834","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":5,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":2,"errorCode":0,"baseOffset":"59","logAppendTime":"1556623780833","logStartOffset":"0"}]}],"throttleTime":0}}
{"logger":"kafkajs","broker":"LAPTOP-XXXXXXXX:9092","clientId":"my-app2","correlationId":4,"size":58,"data":{"topics":[{"topicName":"topic-test","partitions":[{"partition":0,"errorCode":0,"baseOffset":"59","logAppendTime":"1556623780834","logStartOffset":"0"}]}],"throttleTime":0}}
```

Based on this, I'm quite confident that the producing works.

Could you try removing `partitionsConsumedConcurrently` from your consumer? I don't think it's likely, but that is a new code path that could potentially be a source of problems.

**Comment by vykimo** [comment]
Here is my consumer log, without partitionsConsumedConcurrently :
https://pastebin.com/dGupaK5r

**Comment by vykimo** [comment]
Up ! 

I tried this code :
```
    await Promise.all(
        messages.map(message => producer.send({ topic: 'topic-test', messages: [message] }))
    )
```
And I still have the same issue ; with idempotency only 4 messages, without it, I have my 10 msgs.

**Comment by vykimo** [comment]
One thing that I noticed and maye it can help you @Nevon.

Consumed messages have been produced with this info returned :
`      logAppendTime: '-1',`
And skipped messages have this info returned by producer :
`      logAppendTime: '1556789239001',`

And I saw on kafka logs, messages are not produced.

**Comment by Nevon** [comment]
I think I know what the problem might be. The way that Kafka implements idempotency is by each producer having an id and a sequence number. Each request from the producer should have a sequence number that is higher than any previous request. If Kafka receives several requests with the same sequence number, it discards all except the first one. This means that an idempotent producer can retry the same request many times without the messages being duplicated.

The sequence numbers are tracked in the producer state per topic-partition, and is a simple number that is incremented **on any response from the broker** (basically, once we get a response, we know that the request succeeded and can increment the sequence). Since you are firing off several requests for the same partition without waiting for the requests to complete, the sequence number is the same for all the requests, which means that Kafka will just ignore them as duplicates.

I'm not sure if this is something we can fix on the KafkaJS side, but is there a reason you can't just await the `producer.send` call? If you want to produce many messages, like in your example, you can produce them all in a single call anyway.

**Comment by vykimo** [comment]
I see, it was my impression too.
But I cannot await for each message because my software produce message when it receives an API request.
So here are solutions I see :
1. Everytime you want to produce an idempotent message you have to retrieve latest sequence number and upgrade it.
2. Set at-least-once semantics for my problem. Do you have tutorial where this system is described ?

**Comment by Nevon** [comment]
Another option for you would be to create a new producer whenever you receive an API request, since the sequence number is per producer. The point of an idempotent producer is that you can retry *the same request* infinitely and still be sure that you won't get duplicates. If you are actually producing *different* messages and you don't care about the order of the messages, you should probably not use the same producer.

Your second solution is usually preferred, as it is vastly simpler than trying to achieve exactly-once-semantics. To get at least once semantics, all you need to do is configure the producer with `acks=-1` (this is the default value). If the producer fails to send the message it will retry according to your `retry` configuration, and then eventually throw if it fails completely. At that point it will be your application's responsibility to deal with the failure (by retrying at a later point or something).

**Comment by vykimo** [comment]
Oh this is a nice idea.
But tis the creation of a new producer an expensive operation? And will it resist to the load ? (Approx. 3 calls per sec)

**Comment by tulios** [comment]
@vykimo I think you are approaching this problem from the wrong side, if you can't `await` for `producer#send`, you can't really use idempotent producers. The primitive behind idempotent producers is that they will produce in sequence, event the number of in-flight requests have to be controlled for this to work. For your current design, you have to use a regular producer.

My suggestion is: produce with a regular producer, you can always group messages by key, then set up a consumer to consume this topic and produce to another one using the idempotent producer. In the consumer, you can `await` for send, and have a fully functioning idempotent producer.

@Nevon suggestion probably works, but I would say it's a hack. I wouldn't go for that.

**Comment by vykimo** [comment]
Thank you for your suggestion @tulios but my not quite sure about it because the pre-step with regular producer will cancel (or at least minimize) benefits of idempotency on the second producer... (in the case of the regular one sends twice or fails, idempotent producer will propagate errors thereby)

**Comment by vykimo** [comment]
I did some experimentations and here is my solution :
- I calculated that producing mass messages with one single idempotent Producer takes 1ms for each. 
Althought it works with synchronous productions, the idempotency doesn't work with async messages. 

In my case, API requests are processed asynchronously to avoid blocking state if one go wrong.

- The Creation, Connection, Message Production and Disconnection of an idempotent Producer takes only 6ms! Which is way correct. And it supports up to 1000 requests at the same time with no error (I get `[BrokerPool] Failed to connect to seed broker` with 1500 instant requests).

So I'll switch to the second option, with a new Producer created for every async request.
